odl/tomo/data/uncompr_bin.py

```diff
@@ -0,0 +1,716 @@
+
+        Parameters
+        ----------
+        file : file-like or str
+            Stream or filename from which to read the data. The stream
+            is allowed to be already opened in ``'rb'`` mode.
+        header_fields : sequence of dicts, optional
+            Definition of the fields in the header (per row), each
+            containing key-value pairs for the following keys:
+
+            - ``'name'`` : Label for the field.
+            - ``'offset'`` : Start of the field in bytes.
+            - ``'size'`` : Size of the field in bytes.
+            - ``'dtype'`` : Data type in Numpy- or Numpy-readable format.
+            - ``'dshape'`` (optional) : The array of values is reshaped to
+              this shape.
+            - ``'description'`` (optional) : A human-readable description
+              of the field.
+
+            If empty, the header size is set 0, and the data type is
+            assumed to be ``dtype``. Use this in conjunction with
+            parametrized `read_data` to bypass the header and read
+            arbitrary data portions.
+
+        dtype : optional
+            Data type of the file's data block. It must be understood by
+            the `numpy.dtype` constructor. By default, the data type
+            is determined from the file header, or, if no information
+            is available there, it is set to ``np.dtype(float)``.
+
+        See Also
+        --------
+        header_fields_from_table :
+            Function to parse a specification table, returning a field
+            sequence usable as ``header_fields`` parameter.
+        """
+        # Need those attrs in subsequent code
+        file_attrs = ('mode', 'seek', 'read', 'close')
+        is_file = all(hasattr(file, attr) for attr in file_attrs)
+        if is_file:
+            self.__file = file
+            self.__file_shared = True
+        else:
+            self.__file = open(file, 'rb')
+            self.__file_shared = False
+
+        if 'b' not in self.file.mode:
+            raise ValueError("`file` must be opened in binary mode, "
+                             "but mode 'is {}'".format(self.file.mode))
+
+        try:
+            iter(header_fields)
+        except TypeError:
+            raise TypeError('`header_fields` must be iterable, got '
+                            '{!r}'.format(header_fields))
+        self.__header_fields = header_fields
+
+        # Set default values for some attributes
+        self._init_data_dtype = np.dtype(dtype)
+        self.__header = OrderedDict()
+
+    @property
+    def file(self):
+        """File object from which ``self`` reads."""
+        return self.__file
+
+    def __enter__(self):
+        """Initializer for the context manager."""
+        return self
+
+    def __exit__(self, *exc):
+        """Cleanup before on exiting the context manager."""
+        if not self.__file_shared:
+            self.file.close()
+
+    @property
+    def header_size(self):
```

Comment:
I was thinking about that. I'm counting 5 common properties. Does that warrant a base class? Apart from that, they're not that terribly similar. Probably it's better to use a common mixin class instead of a common base.